op-supervisor/supervisor/backend/rewinder/rewinder.go

```diff
@@ -0,0 +1,256 @@
+	if err != nil {
+		return fmt.Errorf("failed to get cross safe L1 block for chain %s: %w", chainID, err)
+	}
+
+	// If we're still on the canonical chain, nothing to do
+	if crossSafeL1.Hash == newTip.Hash {
+		return nil
+	}
+
+	// Get the finalized block as our lower bound
+	finalized, err := r.db.Finalized(chainID)
+	if err != nil {
+		return fmt.Errorf("failed to get finalized block for chain %s: %w", chainID, err)
+	}
+	finalizedL1, err := r.db.CrossDerivedFromBlockRef(chainID, finalized.ID())
+	if err != nil {
+		return fmt.Errorf("failed to get finalized L1 block for chain %s: %w", chainID, err)
+	}
+
+	// Find the common ancestor by walking back through L1 blocks
+	commonL1Ancestor := finalizedL1.ID()
+	currentL1 := crossSafeL1.ID()
+	for currentL1.Number >= finalizedL1.Number {
+		// Get the canonical L1 block at this height from the node
+		remoteL1, err := r.l1Node.L1BlockRefByNumber(context.Background(), currentL1.Number)
+		if err != nil {
+			return fmt.Errorf("failed to get L1 block at height %d: %w", currentL1.Number, err)
+		}
+
+		// If hashes match, we found the common ancestor
+		if remoteL1.Hash == currentL1.Hash {
+			commonL1Ancestor = currentL1
+			break
+		}
+
+		// Get the previous L1 block from our DB
+		prevDerivedFrom, err := r.db.PreviousDerivedFrom(chainID, currentL1)
+		if err != nil {
+			// If we hit the first block, use it as common ancestor
+			if errors.Is(err, types.ErrPreviousToFirst) {
+				// Still need to verify this block is canonical
+				remoteFirst, err := r.l1Node.L1BlockRefByNumber(context.Background(), currentL1.Number)
+				if err != nil {
+					return fmt.Errorf("failed to get first L1 block: %w", err)
+				}
+				if remoteFirst.Hash == currentL1.Hash {
+					commonL1Ancestor = currentL1
+				} else {
+					// First block isn't canonical, use finalized
+					commonL1Ancestor = finalizedL1.ID()
+				}
+				break
+			}
+			return fmt.Errorf("failed to get previous L1 block: %w", err)
+		}
+
+		// Move to the parent
+		currentL1 = prevDerivedFrom.ID()
+	}
+
+	// Get the last L2 block derived from the common ancestor
+	crossSafeDerived, err := r.db.LastDerivedFrom(chainID, commonL1Ancestor)
+	if err != nil {
+		return fmt.Errorf("failed to get derived from for chain %s: %w", chainID, err)
+	}
+
+	// Rewind CrossSafe to the derived block
+	if err := r.db.RewindCrossSafe(chainID, crossSafeDerived); err != nil {
+		return fmt.Errorf("failed to rewind cross-safe for chain %s: %w", chainID, err)
+	}
+
+	// Now check LocalSafe DB for reorg
+	localSafeDerived, err := r.checkLocalSafeForReorg(chainID, crossSafeDerived)
+	if err != nil {
+		return fmt.Errorf("failed to check local-safe for reorg for chain %s: %w", chainID, err)
+	}
+
+	// Rewind LocalSafe to the derived block
+	if err := r.db.RewindLocalSafe(chainID, localSafeDerived); err != nil {
+		return fmt.Errorf("failed to rewind local-safe for chain %s: %w", chainID, err)
+	}
```

Comment:
As mentioned above, I *think* you want to switch your use of local and cross safe in this function -- the reorg ancestor test should be conducted with the local safe db, since it has more data and is "further ahead". the cross-safe database may not need to be rewound if it is far enough behind.